service/controller/resource/azureconfigfinalizer/resource.go

```diff
@@ -0,0 +1,103 @@
+package azureconfigfinalizer
+
+import (
+	"context"
+	"strings"
+
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/azure-operator/v4/service/controller/key"
+)
+
+const (
+	// Name is the identifier of the resource.
+	Name = "azureconfigfinalizer"
+
+	// Finalizer of old operator's controller.
+	legacyFinalizer = "operatorkit.giantswarm.io/azure-operator"
+)
+
+type Config struct {
+	CtrlClient client.Client
+	Logger     micrologger.Logger
+}
+
+// Resource manages the cluster ID label in the AzureConfig CR.
+type Resource struct {
+	ctrlClient client.Client
+	logger     micrologger.Logger
+}
+
+func New(config Config) (*Resource, error) {
+	if config.CtrlClient == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.CtrlClient must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	r := &Resource{
+		ctrlClient: config.CtrlClient,
+		logger:     config.Logger,
+	}
+
+	return r, nil
+}
+
+// EnsureCreated ensures that reconciled AzureConfig CR gets orphaned finalizer
+// deleted.
+func (r *Resource) EnsureCreated(ctx context.Context, obj interface{}) error {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring azureconfig doesn't have orphaned azure-operator finalizer present")
+
+	{
+		// Refresh the CR object.
+		err := r.ctrlClient.Get(ctx, client.ObjectKey{cr.Name, cr.Namespace}, &cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var exists bool
+	for i, v := range cr.Finalizers {
+		if strings.TrimSpace(v) == legacyFinalizer {
+			exists = true
+
+			// Drop it.
+			cr.Finalizers = append(cr.Finalizers[:i], cr.Finalizers[i+1:]...)
+			break
```

Comment:
Instead of break here, you could update the CR straight away and return I guess.
But not important

Reply:
Sure, but I feel the loop would get a bit too messy with all the logging & stuff. I prefer it this way. :slightly_smiling_face: